### ISSUE TITLE
kubectl diff fails to detect deletions dade in danifest after kubectl create

### ISSUE THREAD

**Issue by hlastras** [kubernetes/kubectl#1587]
**What happened**:
When using `kubectl create` to deploy resources and subsequently modifying the original manifest (specifically deleting an environment variable), `kubectl diff` does not detect the deletion, although it correctly detects changes (e.g., updating an environment variable value). This behavior contrasts with `kubectl apply`, where such deletions are detected.

**What you expected to happen**:
I expected kubectl diff to show all differences between the current state of the cluster and the state defined by the modified manifest, including deletions of elements such as environment variables.

**How to reproduce it (as minimally and precisely as possible)**:
```bash
cat << EOF > deployment.yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: busybox-deployment
  labels:
    app: busybox
spec:
  replicas: 0
  selector:
    matchLabels:
      app: busybox
  template:
    metadata:
      labels:
        app: busybox
    spec:
      containers:
      - name: busybox
        image: busybox
        env:
        - name: TESTENV1
          value: "var1"
        - name: TESTENV2
          value: "var2"
EOF

kubectl create -f deployment.yaml

# Apply some changes on the file (deleting an envvar and updating another)
sed -i '/- name: TESTENV1/{N;d;}' deployment.yaml
sed -i 's/value: "var2"/value: "foo"/' deployment.yaml

kubectl diff -f deployment.yaml
```

**Output**:
```diff
@@ -4,7 +4,7 @@
   annotations:
     deployment.kubernetes.io/revision: "1"
   creationTimestamp: "2024-04-22T21:04:20Z"
-  generation: 1
+  generation: 2
   labels:
     app: busybox
   name: busybox-deployment
@@ -34,7 +34,7 @@
         - name: TESTENV1
           value: var1
         - name: TESTENV2
-          value: var2
+          value: foo
         image: busybox
         imagePullPolicy: Always
         name: busybox
```

**Expected output**:
```diff
@@ -6,7 +6,7 @@
   annotations:
     deployment.kubernetes.io/revision: "1"
   creationTimestamp: "2024-04-22T21:04:28Z"
-  generation: 1
+  generation: 2
   labels:
     app: busybox
   name: busybox-deployment
@@ -33,10 +33,8 @@
     spec:
       containers:
       - env:
-        - name: TESTENV1
-          value: var1
         - name: TESTENV2
-          value: var2
+          value: foo
         image: busybox
         imagePullPolicy: Always
         name: busybox
```
If on the above script I change `kubectl create` by `kubectl apply`, the output is as expected.

**Anything else we need to know?**:
This behavior has been consistently reproducible across different environments, suggesting it might be inherent to the current implementation of `kubectl diff` rather than an environment-specific issue. Apologies in advance if this is the expected behavior, but I could not find any clear documentation explaining this.

**Environment**:
- Kubernetes client and server versions (use `kubectl version`):
  - AWS EKS cluster version 1.27 with kubectl version 1.27
  - Kind cluster version 1.27 with kubectl version 1.27
  - Kind cluster version 1.29 with kubectl version 1.29
- Cloud provider or hardware configuration: AWS EKS, Kind clusters
- OS: MacOS Monterey 12.6

**Comment by ardaguclu** [comment]
I tested this with the same resource and could have reproduced it when I use `create` command to create the resources.
However, if I follow the same steps with the only difference by using `apply` instead of create, it worked;

```sh
cat << EOF > deployment.yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: busybox-deployment
  labels:
    app: busybox
spec:
  replicas: 0
  selector:
    matchLabels:
      app: busybox
  template:
    metadata:
      labels:
        app: busybox
    spec:
      containers:
      - name: busybox
        image: busybox
        env:
        - name: TESTENV1
          value: "var1"
        - name: TESTENV2
          value: "var2"
EOF

kubectl apply -f deployment.yaml

# Apply some changes on the file (deleting an envvar and updating another)
sed -i '/- name: TESTENV1/{N;d;}' deployment.yaml
sed -i 's/value: "var2"/value: "foo"/' deployment.yaml

kubectl diff -f deployment.yaml
```

```diff
 creationTimestamp: "2024-04-26T10:13:54Z"
-  generation: 1
+  generation: 2
   labels:
     app: busybox
   name: busybox-deployment
@@ -33,10 +33,8 @@
     spec:
       containers:
       - env:
-        - name: TESTENV1
-          value: var1
         - name: TESTENV2
-          value: var2
+          value: foo
         image: busybox
         imagePullPolicy: Always
         name: busybox
```

@hlastras As a sig-cli maintainers, we always recommend to use `apply` command rather than `create`. Apply is declarative way of managing resources and always should be the preferred approach. 

I'll not close this issue because apparently there is mismatch between create and diff commands (diff command works more aligning with apply) in case someone would want to spend time on it.
/unassign

**Comment by Coen90** [comment]
Hi @ardaguclu 
I would like to work on this issue.

**Comment by lbcd** [comment]
Any news? Thanks for working on :)

**Comment by jsolana** [comment]
Any news? Thanks!

**Comment by HontoNoRoger** [comment]
This misbehavior seems to be a rather severe bug, as automated continuous deployment tools such as https://github.com/argoproj/argo-cd/issues/4537 depend on `kubectl diff` and `kubectl apply` for the configuration drift detection.

As a result (as also presented in the issue ticket) configuration drift (by accident, malicious intend or whatnot) is not corrected.

**Comment by brianpursley** [comment]
This is confusing because I think there might be an actual bug here (see bottom of this comment), but it mostly has to do with differences between client-side apply and server-side apply, and how mixing the two can result in unexpected output.

| Command | Field Management | Diff Command Needed | Works? |
| --- | --- | --- | --- |
| `kubectl apply -f deployment.yaml` | Client-Side | `kubectk diff -f deployment.yaml` | ✅ |
| `kubectl create -f deployment.yaml --save-config` | Client-Side | `kubectl diff -f deployment.yaml` | ✅ |
| `kubectl apply -f deployment.yaml --server-side` | Server-Side | `kubectl diff -f deployment.yaml --server-side` | ✅  |
| `kubectl create -f deployment.yaml` | Server-Side | `kubectl diff -f deployment.yaml --server-side` | ❌ |

* Client-Side = fields managed using the `kubectl.kubernetes.io/last-applied-configuration` annotation.
* Server-Side = fields managed using the `managedFields` metadata property.

Your diff is going to possibly be incorrect if you mix-and-match client-side-apply and server-side-apply, particularly when it comes to comparing lists where an item is removed, like in the case of environment variables.

---

POTENTIAL BUG: It seems like I *should* be able to create the deployment without `--save-config` and then use `kubectl diff -f deployment.yaml --server-side` successfully, but that's not the case.

I get this error when attempting the diff using `--server-side`
```
$ kubectl diff -f deployment.yaml --server-side
W0601 15:46:15.285462    3459 diff.go:721] Object (apps/v1, Kind=Deployment: busybox-deployment) keeps changing, diffing without lock
Error from server (Conflict): Apply failed with 1 conflict: conflict with "kubectl-create" using apps/v1: .spec.template.spec.containers[name="busybox"].env[name="TESTENV2"].value
deployment.apps "busybox-deployment" deleted from default namespace
```

And if I use `--force-conflicts`, it fails to detect the removed environment variable:
```
$ kubectl diff -f deployment.yaml --server-side --force-conflicts
diff -u -N /var/folders/vq/j16ggg_55vjg9j9g19blcs580000gn/T/LIVE-2205991979/apps.v1.Deployment.default.busybox-deployment /var/folders/vq/j16ggg_55vjg9j9g19blcs580000gn/T/MERGED-1780882091/apps.v1.Deployment.default.busybox-deployment
--- /var/folders/vq/j16ggg_55vjg9j9g19blcs580000gn/T/LIVE-2205991979/apps.v1.Deployment.default.busybox-deployment	2026-06-01 15:46:53
+++ /var/folders/vq/j16ggg_55vjg9j9g19blcs580000gn/T/MERGED-1780882091/apps.v1.Deployment.default.busybox-deployment	2026-06-01 15:46:53
@@ -4,7 +4,7 @@
   annotations:
     deployment.kubernetes.io/revision: "1"
   creationTimestamp: "2026-06-01T19:46:53Z"
-  generation: 1
+  generation: 2
   labels:
     app: busybox
   name: busybox-deployment
@@ -34,7 +34,7 @@
         - name: TESTENV1
           value: var1
         - name: TESTENV2
-          value: var2
+          value: foo
         image: busybox
         imagePullPolicy: Always
         name: busybox
deployment.apps "busybox-deployment" deleted from default namespace
```

**Comment by brianpursley** [comment]
I'm going to re-accept this because it still seems like a related problem exists here, even if it is slightly different than the original issue's description.

/triage accepted

**Comment by brianpursley** [comment]
A little more information:

When you create a resource using `kubectl create`, its field manager is:
```json
  "manager": "kubectl-create",
  "operation": "Update",
```

When you attempt `kubectl diff`, kubectl performs a dry-run patch operation under the hood which is an `Apply` operation (not `Update`), so even if you specify `--field-manager=kubectl-create` the underlying `HTTP PATCH` response will still be `409 CONFLICT` because the server uses manager + operation to check for a conflict.

Bottom line: You should use `kubectl apply` instead of `kubectl create`.
